### ISSUE TITLE
without-K ?

### ISSUE THREAD

**Issue by JacquesCarette** [agda/agda-stdlib#135]
Would it be a good idea to add the `--without-K` pragma to as much of the library as possible?
  - Yes: then I would start submitting pull requests (on an ad hoc basis) doing this.
  - No: why not?  And where would this rationale be best documented?  [I can take care of taking the discussion and adding it to the docs]

**Comment by MatthewDaggitt** [comment]
I'm probably not the best person to answer this as I don't fully understand the subtleties of K-axiom. I'm happy to defer to someone who knows more?

@JacquesCarette what would be the advantages of such a change? Would it simply highlight the bits of the library that does rely on the K-axiom? Or does it interact with options somehow?

**Comment by JacquesCarette** [comment]
The main reason would be to clearly label which parts of the standard library can be used as-is in HoTT, which is the main side-effect of highlighting those bits which do not depend on K.

Note that most of the pieces of the standard library that I have used (which is only part!) would go through without any changes.  Understanding what fails might be quite instructive too.

**Comment by MatthewDaggitt** [comment]
It sounds reasonable, but (from a completely non-theoretical perspective!) I'm a little wary of opening the gates to including pragmas at the top of every file. If you include the pragma, does it check subsequent imported files as well? If it does, would it be possible to find the parts of the standard library that use the K-axiom by simply adding it at the top of the `Everything.agda` file?

@andreasabel do you have any thoughts on this?

**Comment by gallais** [comment]
> If you include the pragma, does it check subsequent imported files as well?

It should but it doesn't at the moment. Cf. agda/agda#2487

**Comment by JacquesCarette** [comment]
That bug is definitely a pre-condition to having this issue 'work properly'.  It does not prevent the actual change (adding --without-K, if desired) from being implemented, just its effectiveness.

**Comment by MatthewDaggitt** [comment]
As you say, I'm imagining that large swathes of the library will be fine and therefore adding `--without-K` to every file seems redundant. 

I suggest that we wait until the issue with pragma propagation is resolved, and then revisit this when we can stick the pragma in `Everything.agda` and see what breaks?

**Comment by gallais** [comment]
Command line arguments on the other hand are respected when checking the dependencies of a module. So `agda --without-K -i. -i src/ Everything.agda` will indeed fail whenever there is a K-related issue. E.g. I get:

```
/home/gallais/projects/agda-stdlib/src/Relation/Binary/HeterogeneousEquality.agda:56,9-13
I'm not sure if there should be a case for the constructor refl,
because I get stuck when trying to solve the following unification
problems (inferred index ≟ expected index):
  {_} ≟ {_}
  x ≟ y
Possible reason why unification failed:
  Cannot eliminate reflexive equation A = A of type Set a because K
  has been disabled.
when checking that the pattern refl has type x ≅ y
```

**Comment by JacquesCarette** [comment]
Ah, that's good to know (about command-line).  It is unsurprising that heterogeneous equality uses K.  It would be interesting to see if it is fundamental -- it might be.  @jespercockx might know.

**Comment by andreasabel** [comment]
Currently, if we put `--without-K` at the beginning of a file, it does not prevent import of modules that use K.  Thus, at the current stage, such an information might be not very helpful and could even be misleading (e.g. if that suggests to you you can use this module of the standard-library in a K-free type theory).

Thus, before we move on here, we should be clear what the aims of putting --without-K into the standard library should be.

**Comment by JacquesCarette** [comment]
@andreasabel : agreed on waiting.

I think the aim is clear: to be able to safely re-use as much of the standard library in a HoTT / raw MLTT setting as possible.  I have had to duplicate parts of the library, just to make sure it was K-free.  It was a trivial exercise (as it was indeed largely K-free), but tedious.

**Comment by asr** [comment]
> raw MLTT

:+1:

**Comment by jespercockx** [comment]
> Ah, that's good to know (about command-line). It is unsurprising that heterogeneous equality uses K. It would be interesting to see if it is fundamental -- it might be.

The eliminator for heterogeneous equality given by Conor McBride in his thesis is equivalent with K, so this is indeed fundamental. Without K, you can still define (Conor's) heterogeneous equality but is is quite useless without this eliminator. Instead, you can use the 'path over' version of heterogeneous equality from HoTT.

**Comment by jespercockx** [comment]
Back in 2014, I did some experiments with adding `--without-K` to all files in the standard library. I also kept some notes on the problems I encountered, I'm including them below. It's possible that some of these issues have gone away in the meantime because of improvements to the pattern matching unifier.

```
MODULE                                      DEFINITION		    AT WHAT TYPE?       NB OF ISSUES   
==================================================================================================

Relation.Binary.HeterogeneousEquality.Core  ≅-to-≡                  Set                 28
| Relation.Binary.PropositionalEquality     _≅⟨_⟩_ 
| Relation.Binary.HeterogeneousEquality     ≡↔≅
| Relation.Binary.HeterogeneousEquality     ≡-ext-to-≅-ext
| Relation.Binary.Sigma.Pointwise           ⇔ 
| Relation.Binary.Sigma.Pointwise           ⇔-↠
| Relation.Binary.Sigma.Pointwise           ↣
| Relation.Binary.Sigma.Pointwise           ↞
| Relation.Binary.Sigma.Pointwise           ↠
| Relation.Binary.Sigma.Pointwise           ↔
| Relation.Binary.Sigma.Pointwise           cong
| Data.Container.Any                        cong
| Data.Container.Any                        swap
| Data.Container.Any                        remove-linear
| Data.Container.Any                        linear-identity
| Data.Container.Any                        join↔◇
| Data.List.Any.Properties                  Any-cong
| Data.List.Any.Properties                  swap
| Data.List.Any.Properties                  ⊛↔
| Data.List.Any.Properties                  ⊛⁺′
| Data.List.Any.Properties                  ⊗↔
| Data.List.Any.Properties                  ⊗↔′
| Data.List.Any.BagAndSetEquality           >>=-left-distributive
| Data.List.Any.BagAndSetEquality           ⊛-left-distributive
| Data.List.Any.Membership                  concat-∈↔
| Data.List.Any.Membership                  ⊛-∈↔
| Data.List.Any.Membership                  concat-mono
| Data.List.Any.Membership                  _⊛-mono_

Relation.Binary.PropositionalEquality       proof-irrelevance       typevar             10
| Data.Vec.Properties                       []=↔lookup
| Data.Container.Any                        map-cong
| Data.List.Any.BagAndSetEquality           map-cong
| Data.List.Any.BagAndSetEquality           concat-cong
| Data.List.Any.BagAndSetEquality           >>=-cong
| Data.List.Any.BagAndSetEquality           ⊛-cong
| Data.List.Any.BagAndSetEquality           ⊗-cong
| Data.List.Any.Membership                  ⊗-∈↔
| Data.List.Any.Membership                  _⊗-mono_


Algebra.RingSolver                          _≟H_                    Nat                 fix 
Algebra.RingSolver                          _≟N_                    Nat                 fix

Data.Fin.Properties                         drop-suc                Nat                 fix 

Data.Vec.Equality                           trans                   Nat                 fix
Data.Vec.Equality                           _≟_                     Nat                 fix

Relation.Binary.HeterogeneousEquality       subst                   Set                 3
| Relation.Binary.HeterogeneousEquality     subst-removable 
| Relation.Binary.HeterogeneousEquality     resp₂
Relation.Binary.HeterogeneousEquality       subst₂                  Set                 1
Relation.Binary.HeterogeneousEquality       cong                    Set                 3
| Relation.Binary.HeterogeneousEquality     cong-app
| Data.Container.Indexed                    Eq⇒≅
Relation.Binary.HeterogeneousEquality       cong₂                   Set                 1
Relation.Binary.HeterogeneousEquality       proof-irrelevance       Set                 1

Data.Vec.Properties                         ∷-injective             Nat                 fix
Data.Vec.Properties                         proof-irrelevance-[]=   typevar             ???
| Data.Vec.Properties                       []=↔lookup              
Data.Vec.Properties                         []≔-minimal             Nat                 fix

Relation.Binary.Vec.Pointwise               head                    Nat                 fix
Relation.Binary.Vec.Pointwise               tail                    Nat                 fix

Data.Fin.Subset.Properties                  drop-there              Nat                 fix
Data.Fin.Subset.Properties                  ∉⊥                      Nat                 fix
Data.Fin.Subset.Properties                  x∈⁅y⁆⇔x≡y               Nat                 fix
Data.Fin.Subset.Properties                  ∪⇔⊎                                         Nat                 fix
Data.Fin.Subset.Properties                  orders-equivalent       Nat                 fix

Data.Fin.Dec                                _∈?_                    Nat                 fix

Data.Colist                                 Any-cong                typevar             5
| Data.Colist.Infinite-merge                Any-⋎P              
| Data.Colist.Infinite-merge                index-Any-⋎P
| Data.Colist.Infinite-merge                Any-merge
| Data.Colist.Infinite-merge                ∈-merge
Data.Colist                                 ⊑-Poset                 typevar             4
| Data.Colist                               ⊑-Reasoning  
| Data.Colist                               ⊆-Preorder  
| Data.Colist                               ⊆-Reasoning 

Relation.Binary.Sigma.Pointwise             Rel↔≡                   Set                 1
Relation.Binary.Sigma.Pointwise             inverse                 typevar             1

Data.Container.Indexed                      setoid                  Set                 3
| Data.Container.Indexed                    identity        
| Data.Container.Indexed                    composition     
Data.Container.Indexed                      natural                 Set                 1  
Data.Container.Indexed                      ∘-correct               Set                 1 

Data.Covec                                  setoid                  Conat               2
| Data.Covec                                poset                

Data.List.Any.BagAndSetEquality             drop-cons               typevar             1

Data.List.Countdown                         drop-suc                Nat                 fix

Data.Star.Decoration                        gmapAll                 typevar             1
Data.Star.Decoration                        _◅◅◅_                   typevar             3
| Data.Star.Decoration                      _▻▻▻_
| Data.Star.Vec                             _++_

Data.Star.Pointer                           lookup                  typevar             4
| Data.Star.Pointer                         last    
| Data.Star.Environment                     lookup
| Data.Star.Vec                             lookup

```

**Comment by JacquesCarette** [comment]
Thanks.  I am not surprised that the original eliminator is in fact equivalent to K.  And indeed, HoTT is part of the motivation for this.

Thanks a lot for the detailed summary of the experiments.  I do believe that quite a few of these should indeed be fixed, as there is no fundamental reason to use K.  Some of these are quite intriguing - the ones in Data.List.Any in particular jump out at me.  I would be surprised if they were essential.

**Comment by MatthewDaggitt** [comment]
> Some of these are quite intriguing - the ones in Data.List.Any in particular jump out at me. I would be surprised if they were essential.

So @nad was trying to some pretty advanced stuff in `Data.List.Any` that's not found anywhere else in the library. In particular proving relationships about general notions of equivalences. I'm currently grappling with trying to refactor some it (#155) at the moment which might fix some of the problems.

**Comment by MatthewDaggitt** [comment]
Actually I find it a bit disturbing that all the `drop` lemmas aren't compatible with it. They're pretty vital...

**Comment by JacquesCarette** [comment]
I am actually quite aware of @nad's work on general notions of equivalences -- read the paper, enjoyed it, using it / generalizing it now.  I am trying to implement a notion of `Multiset` which does not require decidable equality; turns out that work on equivalences is one key part to make it 'work'.  Another seems to be the notion of SetoidFamily (a la P. Dybjer).

I would be quite interested in version of `Data.List.Any` which didn't assume propositional equality, but worked over any `Setoid`.  Not having that is what set me on the path of SetoidFamily!

**Comment by nad** [comment]
I want to be able to depend on parts of the standard library in code that uses `--without-K`, so I have started work on this issue (but I don't know if or when I'll get around to finishing it). I've added an option `--without-K` to most modules. Exceptions:
* The modules dealing with heterogeneous equality.
* Modules that depend on `Agda.Builtin.TrustMe` or `Agda.Builtin.Equality.Erase`.
* Perhaps something else that I've forgotten now.

Some of the remaining code uses the K rule in an inessential way, and can be rewritten. Other definitions (like some of "the `drop` lemmas") cannot be fixed. I suggest that such code is moved from module *M* to module *M*`.K`.

**Comment by jespercockx** [comment]
Maybe we could name the module `M.UIP` or `M.With-K` instead of just `M.K`?

**Comment by nad** [comment]
I'm going with `With-K`, which matches the name of the option.

**Comment by MatthewDaggitt** [comment]
Sorry missed this. Can we go with "WithK" instead? I've spent the last few
releases removing inconsistent hyphens from module names!

**Comment by nad** [comment]
I would have preferred to abandon camel case, but I can see the appeal of being consistent.

**Comment by nad** [comment]
I pushed a fix now. Perhaps it would have been better to first submit a pull request. However, currently Agda does not check that modules that use `--without-K` do not import modules that use `--with-K`, and there are still library modules that use `--with-K`, so I did not want to sit on this for too long. I hope that others who update the library now take care to not import modules that use `--with-K` from modules that use `--without-K`.

Some things to consider:
* I added a module `Function.HalfAdjointEquivalence`. This module does not use setoids, unlike other code under `Function`.
* I added some equality rearrangement lemmas to `Relation.Binary.PropositionalEquality`. Perhaps it would make sense to put such lemmas in a different module.
* The new proof of `Data.List.Relation.BagAndSetEquality.drop-cons` includes a number of private lemmas, some of which could perhaps be exported.
* Some code in `Data.Product.Relation.Pointwise.Dependent` no longer depends on the type families defined at the start of that module. Perhaps that code could be moved somewhere else.
* In order to make it easier to prove properties about equality proofs defined using equational reasoning I suggest that `Relation.Binary.PropositionalEquality.Core.trans` is changed so that it matches on the second argument instead of the first.

**Comment by nad** [comment]
Currently the following modules use `--with-K`:

* `Data.AVL.Indexed.WithK`.
* `Data.Char.Unsafe`.
* `Data.Container.Indexed.WithK`.
* `Data.Float.Unsafe`.
* `Data.Nat.DivMod.Unsafe`.
* `Data.Nat.Unsafe`.
* `Data.Product.Properties.WithK`.
* `Data.Product.Relation.Pointwise.Dependent.WithK`.
* `Data.Star.BoundedVec`.
* `Data.Star.Decoration`.
* `Data.Star.Environment`.
* `Data.Star.Fin`.
* `Data.Star.Pointer`.
* `Data.Star.Vec`.
* `Data.String.Unsafe`.
* `Data.Vec.Properties.WithK`.
* `Data.Vec.Relation.Equality.Propositional.WithK`.
* `Data.W.WithK`.
* `Data.Word.Unsafe`.
* `Reflection`.
* `Relation.Binary.Construct.Closure.ReflexiveTransitive.Properties.WithK`.
* `Relation.Binary.Construct.Closure.Transitive.WithK`.
* `Relation.Binary.HeterogeneousEquality`.
* `Relation.Binary.HeterogeneousEquality.Core`.
* `Relation.Binary.HeterogeneousEquality.Quotients`.
* `Relation.Binary.HeterogeneousEquality.Quotients.Examples`.
* `Relation.Binary.PropositionalEquality.TrustMe`.
* `Relation.Binary.PropositionalEquality.WithK`.

**Comment by nad** [comment]
> * I added some equality rearrangement lemmas to `Relation.Binary.PropositionalEquality`. Perhaps it would make sense to put such lemmas in a different module.

I only included lemmas that I used. One could add many more. A [solver](http://www.cse.chalmers.se/~nad/listings/equality/Equality.Tactic.html) for equalities between equalities might also be nice.

**Comment by MatthewDaggitt** [comment]
> Perhaps it would have been better to first submit a pull request.

Yes please do this in future! This is a fairly major rearrangement and it would nice to have allowed a chance for people to look through it.

**Comment by nad** [comment]
The problem was that for every commit that came in I had to check manually whether this commit broke anything. Hopefully https://github.com/agda/agda/issues/2487 will be fixed soon.

**Comment by gallais** [comment]
Is it just me or should `_∈_` defined in [`Data.Container.Indexed.WithK`](https://agda.github.io/agda-stdlib/Data.Container.Indexed.WithK.html#6613) be transferred
to `Data.Container.Indexed`? It should typecheck without K AFAICT.

**Comment by nad** [comment]
It is for some reason defined using heterogeneous equality, so I imagine that it is pretty useless without K. However, you are right that this definition on its own does not depend on the K rule.

**Comment by MatthewDaggitt** [comment]
Fixed with 2811a9b5989c5f06abfa86adba9696880d943583

**Comment by MatthewDaggitt** [comment]
Fixed with 3b4d171

**Comment by nad** [comment]
@MatthewDaggitt, `Relation.Binary.HeterogeneousEquality` does make use of the K rule.

> However, you are right that this definition on its own does not depend on the K rule.

Except that heterogeneous equality violates the new constraints, only enforced when the K rule is off, for the target universe of an inductive family.
